### PR TITLE
Fix "openssl without EC" make error (SASL)

### PR DESCRIPTION
**Pull request by michaelortmann**
Found by: michaelortmann
Patch by: michaelortmann
Fixes: 

One-line summary:
Fix for OpenSSL without EC (SunOS 5.11 Solaris 11.3 I love you Oracle) and ssl_info type (Sun C 5.13 warning).

Additional description (if needed):
Fedora 10 is also fixed by this patch.
CentOS 5.11 may also be fixed by this patch.
```
$ man openssl
[...]
       The  following  features  are  omitted  from  the  binaries  for issues
       including but not limited to patents, trademark, and US export restric-
       tions:  ECC, IDEA, MDC2, RC3,  RC5, 4758_CCA Engine, AEP Engine, Atalla
       Engine, CHIL  Engine,  CSWIFT  Engine,  GMP   Engine,   NURON   Engine,
       PadLock Engine, Sureware Engine, and UBSEC Engine.
[...]
```
Please **misc/runautotools**

Test cases demonstrating functionality (if applicable):
**1. SunOS**
```
cc -g -I.. -I..  -DHAVE_CONFIG_H -I/usr/include -g3 -DDEBUG -DDEBUG_ASSERT -DDEBUG_MEM -DDEBUG_DNS  -c tls.c
"tls.c", line 856: warning: argument #2 is incompatible with prototype:
        prototype: pointer to function(pointer to const struct ssl_st {int version, int type, pointer to const struct ssl_method_st {..} method, pointer to struct bio_st {..} rbio, pointer to struct bio_st {..} wbio, pointer to struct bio_st {..} bbio, int rwstate, int in_handshake, pointer to function(..) returning int handshake_func, int server, int new_session, int quiet_shutdown, int shutdown, int state, int rstate, pointer to struct buf_mem_st {..} init_buf, pointer to void init_msg, int init_num, int init_off, pointer to unsigned char packet, unsigned int packet_length, pointer to struct ssl2_state_st {..} s2, pointer to struct ssl3_state_st {..} s3, pointer to struct dtls1_state_st {..} d1, int read_ahead, pointer to function(..) returning void msg_callback, pointer to void msg_callback_arg, int hit, pointer to struct X509_VERIFY_PARAM_st {..} param, pointer to struct stack_st_SSL_CIPHER {..} cipher_list, pointer to struct stack_st_SSL_CIPHER {..} cipher_list_by_id, int mac_flags, pointer to struct evp_cipher_ctx_st {..} enc_read_ctx, pointer to struct env_md_ctx_st {..} read_hash, pointer to struct comp_ctx_st {..} expand, pointer to struct evp_cipher_ctx_st {..} enc_write_ctx, pointer to struct env_md_ctx_st {..} write_hash, pointer to struct comp_ctx_st {..} compress, pointer to struct cert_st {..} cert, unsigned int sid_ctx_length, array[32] of unsigned char sid_ctx, pointer to struct ssl_session_st {..} session, pointer to function(..) returning int generate_session_id, int verify_mode, pointer to function(..) returning int verify_callback, pointer to function(..) returning void info_callback, int error, int error_code, pointer to function(..) returning unsigned int psk_client_callback, pointer to function(..) returning unsigned int psk_server_callback, pointer to struct ssl_ctx_st {..} ctx, int debug, long verify_result, struct crypto_ex_data_st {..} ex_data, pointer to struct stack_st_X509_NAME {..} client_CA, int references, unsigned long options, unsigned long mode, long max_cert_list, int first_packet, int client_version, unsigned int max_send_fragment, pointer to function(..) returning void tlsext_debug_cb, pointer to void tlsext_debug_arg, pointer to char tlsext_hostname, int servername_done, int tlsext_status_type, int tlsext_status_expected, pointer to struct stack_st_OCSP_RESPID {..} tlsext_ocsp_ids, pointer to struct stack_st_X509_EXTENSION {..} tlsext_ocsp_exts, pointer to unsigned char tlsext_ocsp_resp, int tlsext_ocsp_resplen, int tlsext_ticket_expected, pointer to void tlsext_opaque_prf_input, unsigned int tlsext_opaque_prf_input_len, pointer to struct tls_session_ticket_ext_st {..} tlsext_session_ticket, pointer to function(..) returning int tls_session_ticket_ext_cb, pointer to void tls_session_ticket_ext_cb_arg, pointer to function(..) returning int tls_session_secret_cb, pointer to void tls_session_secret_cb_arg, pointer to struct ssl_ctx_st {..} initial_ctx, pointer to unsigned char next_proto_negotiated, unsigned char next_proto_negotiated_len, pointer to struct stack_st_SRTP_PROTECTION_PROFILE {..} srtp_profiles, pointer to struct srtp_protection_profile_st {..} srtp_profile, unsigned int tlsext_heartbeat, unsigned int tlsext_hb_pending, unsigned int tlsext_hb_seq, int renegotiate, struct srp_ctx_st {..} srp_ctx}, int, int) returning void : "/usr/include/openssl/ssl.h", line 2123
        argument : pointer to void
[...]
cc -KPIC -g -I. -I../../.. -I../../.. -I../../../src/mod  -DHAVE_CONFIG_H -I/usr/include -g3 -DDEBUG -DDEBUG_ASSERT -DDEBUG_MEM -DDEBUG_DNS  -DMAKING_MODS -c .././server.mod/server.c && mv -f server.o ../
".././server.mod/servmsg.c", line 1162: undefined symbol: EC_KEY
".././server.mod/servmsg.c", line 1162: undefined symbol: eckey
".././server.mod/servmsg.c", line 1223: warning: implicit function declaration: EVP_PKEY_get1_EC_KEY
".././server.mod/servmsg.c", line 1230: warning: implicit function declaration: ECDSA_size
".././server.mod/servmsg.c", line 1231: warning: implicit function declaration: ECDSA_sign
".././server.mod/servmsg.c", line 1232: warning: implicit function declaration: EC_KEY_free
cc: acomp failed for .././server.mod/server.c
gmake[2]: *** [../server.o] Error 2
gmake[2]: Leaving directory `/export/home/eugene/usr/src/eggdrop/src/mod/server.mod'
gmake[1]: *** [server.mod_so] Error 2
gmake[1]: Leaving directory `/export/home/eugene/usr/src/eggdrop/src/mod'
gmake: *** [debug] Error 2
```
**2. Fedora 10**
```
./configure
[...]
Operating System: Linux 2.6.27.5-117.fc10.x86_64
IPv6 Support: yes
Tcl version: 8.5.3
SSL/TLS Support: yes (OpenSSL 0.9.8g 19 Oct 2007)
gcc version 4.3.2 20081105 (Red Hat 4.3.2-7) (GCC)
[...]
make
[...]
gcc -fPIC -g -O2 -pipe -Wall -I. -I../../.. -I../../.. -I../../../src/mod  -DHAVE_CONFIG_H -I/home/michael/opt/tcl-8.5.3/include -g3 -DDEBUG -DDEBUG_ASSERT -DDEBUG_MEM -DDEBUG_DNS  -DMAKING_MODS -c .././server.mod/server.c && mv -f server.o ../
In file included from .././server.mod/server.c:144:
.././server.mod/servmsg.c: In function ‘tryauthenticate’:
.././server.mod/servmsg.c:1237: error: ‘EC_KEY’ undeclared (first use in this function)
.././server.mod/servmsg.c:1237: error: (Each undeclared identifier is reported only once
.././server.mod/servmsg.c:1237: error: for each function it appears in.)
.././server.mod/servmsg.c:1237: error: ‘eckey’ undeclared (first use in this function)
.././server.mod/servmsg.c:1298: warning: implicit declaration of function ‘EVP_PKEY_get1_EC_KEY’
.././server.mod/servmsg.c:1305: warning: implicit declaration of function ‘ECDSA_size’
.././server.mod/servmsg.c:1306: warning: implicit declaration of function ‘ECDSA_sign’
.././server.mod/servmsg.c:1307: warning: implicit declaration of function ‘EC_KEY_free’
make[2]: *** [../server.o] Error 1
make[2]: Leaving directory `/home/michael/usr/src/eggdrop/src/mod/server.mod'
make[1]: *** [server.mod_so] Error 2
make[1]: Leaving directory `/home/michael/usr/src/eggdrop/src/mod'
make: *** [debug] Error 2
```

### CODE REVIEW

**Comment by vanosg** [comment]
build

**Comment by vanosg** [comment]
build

File: src/mod/server.mod/servmsg.c [hunk 1]
```diff
@@ -1155,14 +1155,16 @@ static int tryauthenticate(char *from, char *msg)
   #endif
   unsigned char dst[((MAX((sizeof src), 400) + 2) / 3) << 2] = "";
 #ifdef TLS
```
Comment (vanosg) [inline src/mod/server.mod/servmsg.c]: Does HAVE_EVP_PKEY_GET1_EC_KEY imply TLS? Can we get rid of this double-ifdef?

**Comment by vanosg** [comment]
Discussion on IRC- we'll also add a check at load to see (if sasl is requested) if EC libs are missing and NIST256 is selected as the auth method, and prevent bot from running if so. These log messages will remain in, in case some separate Tcl mechanism is used to add or change sasl auth methods after initial load.

File: src/mod/server.mod/servmsg.c (hunk 1 above)
Comment (vanosg) [reply src/mod/server.mod/servmsg.c]: Removed

**Comment by vanosg** [comment]
with sasl 1 and sasl-mechanism 1:
```
Eggdrop v1.9.0+sasl (C) 1997 Robey Pointer (C) 2010-2018 Eggheads
--- Loading eggdrop v1.9.0+sasl (Sat Dec 28 2019)
Listening for telnet connections on 0.0.0.0:3183 (all).
Module loaded: dns             
Module loaded: channels        
* ERROR: NIST256P functionality missing from OpenSSL libs, please choose a different SASL method
```
Loads fine with sasl 0 and sasl-mechanism 0/2

File: src/mod/server.mod/server.c [hunk 2]
```diff
@@ -2156,6 +2146,19 @@ char *server_start(Function *global_funcs)
     module_undepend(MODULE_NAME);
     return "This module requires Eggdrop 1.8.0 or later.";
   }
+/* Some pre-compiled OpenSSL libs (SunOS, I'm looking at you) omit ECC keys
+   due to export restrictions. This prevents us from being able to use the
+   NIST256P method */
+#ifdef TLS
+#ifndef HAVE_EVP_PKEY_GET1_EC_KEY
+if (sasl) {
```
Comment (michaelortmann) [inline src/mod/server.mod/server.c]: what about:

```
.set sasl
Currently: 0
.set sasl 1
Ok, set.
```

? I think this block of code has to happen for Tcl_TraceVar / TCL_TRACE_WRITES on variable sasl instead.

File: src/mod/server.mod/server.c [hunk 3]
```diff
@@ -2203,6 +2190,16 @@ if (sasl) {
   my_tcl_strings[0].buf = botname;
   add_tcl_strings(my_tcl_strings);
   add_tcl_ints(my_tcl_ints);
+#ifdef TLS
+#ifndef HAVE_EVP_PKEY_GET1_EC_KEY
+if (sasl) {
```
Comment (michaelortmann) [inline src/mod/server.mod/server.c:2298]: Its better now, but my last comment holds, it should be placed into a tcl write trace function on variable sasl.

File: src/mod/server.mod/server.c (hunk 2 above)
Comment (vanosg) [reply src/mod/server.mod/server.c]: We had talked about that earlier- this is why we are leaving the error messages at run-time as well. I think that's sufficient to cover what I consider to be a corner-case. I also am hesitant to kill an already-running bot for a non-terrible reason.

File: src/mod/server.mod/server.h [hunk 4]
```diff
@@ -131,6 +131,12 @@ enum {
   SASL_MECHANISM_NUM
 };
 
-extern char const *SASL_MECHANISMS[];
+/* Available sasl mechanisms. */
+char const *SASL_MECHANISMS[SASL_MECHANISM_NUM] = {
```
Comment (thommey) [inline src/mod/server.mod/server.h:135]: designated initializers are C99, did we decide to go with C99?

Comment (vanosg) [reply src/mod/server.mod/server.h:135]: Indeed we did!

**Pull request merged**